File: src/simulation/breaches.ts
import type { Vec3 } from '../ships/blueprint';
import type { CompartmentState } from './damage';
import { dot, normalize, sub } from './geometry';

const cross = (a: Vec3, b: Vec3): Vec3 => [a[1]*b[2]-a[2]*b[1], a[2]*b[0]-a[0]*b[2], a[0]*b[1]-a[1]*b[0]];
function axes(normal: Vec3) {
  const u = normalize(cross(Math.abs(normal[1]) < .9 ? [0, 1, 0] : [0, 0, 1], normal));
  return [u, cross(normal, u)];
}
/** Bounded deterministic aperture-union estimate: sample only the new opening,
 * then subtract portions already open on the same local face. Damage to an
 * existing hole is credited only when it widens it. */
export function addBreach(state: CompartmentState, position: Vec3, areaM2: number, shellId: number,
  apertureRadiusM = Math.sqrt(areaM2 / Math.PI), normal: Vec3 = [Math.sign(position[0]) || 1, 0, 0], incremental = false): number {
  if (!(areaM2 > 0) || !(apertureRadiusM > 0)) return 0;
  normal = normalize(normal);
  const [u, v] = axes(normal), width = areaM2 / (Math.PI * apertureRadiusM);
  const candidates = incremental ? [] : state.breaches.filter(b => {
    const footprint = b.footprintAreaM2 ?? b.areaM2;
    return Math.abs(dot(normal, b.normal ?? normal)) > .95 && Math.abs(dot(sub(position, b.position), normal)) < .05 &&
      Math.hypot(...sub(position, b.position)) < Math.max(width, apertureRadiusM) + Math.max(b.radiusM, footprint / (Math.PI * b.radiusM));
  });
  let uncovered = 128;
  if (candidates.length) {
    uncovered = 0;
    for (let ring = 0; ring < 8; ring++) for (let sector = 0; sector < 16; sector++) {
      const radius = Math.sqrt((ring + .5) / 8), angle = (sector + .5) * Math.PI / 8;
      const point = position.map((n, i) => n + u[i] * width * radius * Math.cos(angle) + v[i] * apertureRadiusM * radius * Math.sin(angle)) as Vec3;
      const covered = candidates.some(b => {
        const [bu, bv] = axes(b.normal ?? normal), delta = sub(point, b.position);
        const fraction = b.areaM2 / (b.initialAreaM2 ?? b.areaM2);
        const bw = (b.footprintAreaM2 ?? b.areaM2) / (Math.PI * b.radiusM);
        return (dot(delta, bu) / bw) ** 2 + (dot(delta, bv) / b.radiusM) ** 2 <= fraction + 1e-9;
      });
      if (!covered) uncovered++;
    }
  }
  const added = Math.max(0, Math.min(areaM2 * uncovered / 128, 4 - state.breachAreaM2));
  if (added <= 0) return 0;
  const separation = (b: typeof state.breaches[number]) => Math.hypot(...sub(b.position, position));
  const closest = state.breaches.reduce<typeof state.breaches[number] | undefined>((best, b) => !best || separation(b) < separation(best) ? b : best, undefined);
  // Incremental growth widens the opening at its own location; it must not
  // migrate a distant shell hole. At the cap preserve hydraulic area/height and
  // bounded cost, where coverage becomes a nearest-cluster approximation.
  if (closest && ((incremental && separation(closest) < .1) || state.breaches.length >= 64)) {
    const total = closest.areaM2 + added;
    closest.position = closest.position.map((n, i) => (n * closest.areaM2 + position[i] * added) / total) as Vec3;
    closest.areaM2 = total; closest.initialAreaM2 = total;
    closest.footprintAreaM2 = (closest.footprintAreaM2 ?? closest.areaM2 - added) + added;
    closest.radiusM = Math.max(closest.radiusM, apertureRadiusM);
  } else state.breaches.push({ position: [...position], areaM2: added, radiusM: apertureRadiusM, shellId,
    footprintAreaM2: areaM2, initialAreaM2: added, normal: [...normal] });
  state.breachAreaM2 += added;
  return added;
}

File: crates/naval-sim/src/breaches.rs
use crate::{
    damage::{Breach, CompartmentState},
    definition::Vec3,
    geometry::*,
};
use std::f64::consts::PI;
fn axes(normal: Vec3) -> (Vec3, Vec3) {
    let u = normalize(cross(
        if normal[1].abs() < 0.9 {
            [0.0, 1.0, 0.0]
        } else {
            [0.0, 0.0, 1.0]
        },
        normal,
    ));
    (u, cross(normal, u))
}
/// Bounded aperture union, matching the authored simulation's 128 deterministic samples.
pub fn add_breach(
    state: &mut CompartmentState,
    position: Vec3,
    area: f64,
    shell_id: i64,
    radius: Option<f64>,
    normal: Option<Vec3>,
    incremental: bool,
) -> f64 {
    let radius = radius.unwrap_or_else(|| (area / PI).sqrt());
    if area <= 0.0 || radius <= 0.0 || !area.is_finite() || !radius.is_finite() {
        return 0.0;
    }
    let normal =
        normalize(normal.unwrap_or([if position[0] < 0.0 { -1.0 } else { 1.0 }, 0.0, 0.0]));
    let (u, v) = axes(normal);
    let width = area / (PI * radius);
    let candidates: Vec<_> = state
        .breaches
        .iter()
        .filter(|b| {
            !incremental
                && dot(normal, b.normal.unwrap_or(normal)).abs() > 0.95
                && dot(sub(position, b.position), normal).abs() < 0.05
                && length(sub(position, b.position))
                    < width.max(radius)
                        + b.radius_m
                            .max(b.footprint_area_m2.unwrap_or(b.area_m2) / (PI * b.radius_m))
        })
        .collect();
    let mut uncovered = 128;
    if !candidates.is_empty() {
        uncovered = 0;
        for ring in 0..8 {
            for sector in 0..16 {
                let r = ((ring as f64 + 0.5) / 8.0).sqrt();
                let angle = (sector as f64 + 0.5) * PI / 8.0;
                let point = std::array::from_fn(|i| {
                    position[i] + u[i] * width * r * angle.cos() + v[i] * radius * r * angle.sin()
                });
                let covered = candidates.iter().any(|b| {
                    let (bu, bv) = axes(b.normal.unwrap_or(normal));
                    let delta = sub(point, b.position);
                    let fraction = b.area_m2 / b.initial_area_m2.unwrap_or(b.area_m2);
                    let bw = b.footprint_area_m2.unwrap_or(b.area_m2) / (PI * b.radius_m);
                    (dot(delta, bu) / bw).powi(2) + (dot(delta, bv) / b.radius_m).powi(2)
                        <= fraction + 1e-9
                });
                if !covered {
                    uncovered += 1;
                }
            }
        }
    }
    let added = (area * uncovered as f64 / 128.0)
        .min(4.0 - state.breach_area_m2)
        .max(0.0);
    if added <= 0.0 {
        return 0.0;
    }
    let closest = state
        .breaches
        .iter()
        .enumerate()
        .min_by(|(_, a), (_, b)| {
            length(sub(a.position, position)).total_cmp(&length(sub(b.position, position)))
        })
        .map(|(i, _)| i);
    if let Some(i) = closest
        && ((incremental && length(sub(state.breaches[i].position, position)) < 0.1)
            || state.breaches.len() >= 64)
    {
        let b = &mut state.breaches[i];
        let total = b.area_m2 + added;
        b.position =
            std::array::from_fn(|i| (b.position[i] * b.area_m2 + position[i] * added) / total);
        b.footprint_area_m2 = Some(b.footprint_area_m2.unwrap_or(b.area_m2) + added);
        b.area_m2 = total;
        b.initial_area_m2 = Some(total);
        b.radius_m = b.radius_m.max(radius);
    } else {
        state.breaches.push(Breach {
            position,
            area_m2: added,
            radius_m: radius,
            shell_id,
            footprint_area_m2: Some(area),
            initial_area_m2: Some(added),
            normal: Some(normal),
        });
    }
    state.breach_area_m2 += added;
    added
}
